internal/kms/kmip.go

```diff
@@ -0,0 +1,518 @@
+			"get Secret %s/%s: %w", kms.namespace, kms.secretName, err)
+	}
+
+	secret, err := c.CoreV1().Secrets(kms.namespace).Get(context.TODO(),
+		kms.secretName, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to get Secret %s/%s: %w",
+			kms.namespace, kms.secretName, err)
+	}
+
+	config := make(map[string]string)
+	for k, v := range secret.Data {
+		switch k {
+		case kmipClientKey, kmipCLientCert, kmipCACert, kmipUniqueIdentifier:
+			config[k] = string(v)
+		default:
+			return nil, fmt.Errorf("unsupported option for KMS "+
+				"provider %q: %s", kmsTypeKMIP, k)
+		}
+	}
+
+	return config, nil
+}
+
+// connect to the kmip endpoint, perform TLS and KMIP handshakes.
+func (kms *kmipKMS) connect() (*tls.Conn, error) {
+	conn, err := tls.Dial("tcp", kms.endpoint, kms.tlsConfig)
+	if err != nil {
+		return nil, err
+	}
+	if kms.readTimeout != 0 {
+		err = conn.SetReadDeadline(time.Now().Add(time.Second * time.Duration(kms.readTimeout)))
+		if err != nil {
+			return nil, err
+		}
+	}
+	if kms.writeTimeout != 0 {
+		err = conn.SetReadDeadline(time.Now().Add(time.Second * time.Duration(kms.writeTimeout)))
+		if err != nil {
+			return nil, err
+		}
+	}
+	defer func() {
+		if err != nil {
+			conn.Close()
+		}
+	}()
+
+	err = conn.Handshake()
+	if err != nil {
+		return nil, err
+	}
+
+	// KMIP handshake
+	err = kms.discover(conn)
+	if err != nil {
+		return nil, err
+	}
+
+	return conn, nil
+}
+
+// discover performs KMIP handshake.
+func (kms *kmipKMS) discover(conn io.ReadWriter) error {
+	respMsg, decoder, uniqueBatchItemID, err := kms.send(conn,
+		kmip14.OperationDiscoverVersions,
+		kmip.DiscoverVersionsRequestPayload{
+			ProtocolVersion: []kmip.ProtocolVersion{
+				{
+					ProtocolVersionMajor: protocolMajor,
+					ProtocolVersionMinor: protocolMinor,
+				},
+			},
+		})
+	if err != nil {
+		return err
+	}
+
+	batchItem, err := kms.verifyResponse(
+		respMsg,
+		kmip14.OperationDiscoverVersions,
+		uniqueBatchItemID)
+	if err != nil {
+		return err
+	}
+
+	ttlvPayload, ok := batchItem.ResponsePayload.(ttlv.TTLV)
+	if !ok {
+		return errors.New("failed to parse responsePayload")
+	}
+
+	var respDiscoverVersionsPayload kmip.DiscoverVersionsResponsePayload
+	err = decoder.DecodeValue(&respDiscoverVersionsPayload, ttlvPayload)
+	if err != nil {
+		return err
+	}
+
+	if len(respDiscoverVersionsPayload.ProtocolVersion) != 1 {
+		return fmt.Errorf("invalid len of discovered protocol versions %v expected 1",
+			len(respDiscoverVersionsPayload.ProtocolVersion))
+	}
+	pv := respDiscoverVersionsPayload.ProtocolVersion[0]
+	if pv.ProtocolVersionMajor != protocolMajor || pv.ProtocolVersionMinor != protocolMinor {
+		return fmt.Errorf("invalid discovered protocol version %v.%v expected %v.%v",
+			pv.ProtocolVersionMajor, pv.ProtocolVersionMinor, protocolMajor, protocolMinor)
+	}
+
+	return nil
+}
+
+// send sends KMIP operation over tls connection, return response and error.
```

Comment:
return response and error... but there are 4 return values? probably good to explain each of them